File: leetcode/p00387/p00387.go
package p00387

// Time Complexity: O(n)
// Space Complexity: O(n)
func firstUniqChar(s string) int {
	memo := map[rune]int{}

	for _, ch := range s {
		memo[ch]++
	}

	for idx, ch := range s {
		if memo[ch] == 1 {
			return idx
		}
	}

	return -1
}

File: leetcode/p00387/p00387.py
from typing import Counter, OrderedDict

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def firstUniqChar(self, s: str) -> int:
        # Ordered Dict will save the characters it encounters in same sequence as the original string.
        for idx, cnt  in OrderedDict(Counter(s)).items():
            if cnt == 1:
                return s.index(idx)
        
        return -1

if __name__ == '__main__':
    s = Solution()

    assert s.firstUniqChar("leetcode") == 0
    assert s.firstUniqChar("loveleetcode") == 2
    assert s.firstUniqChar("aabb") == -1